include/realizations/catchment/Formulation_Manager.hpp

```diff
@@ -285,16 +290,31 @@ namespace realization {
              * 
              * @return std::string of the output root directory
              */
-            std::string get_output_root() const noexcept {
+            std::string get_output_root() const {
                 const auto output_root = this->tree.get_optional<std::string>("output_root");
                 if (output_root != boost::none && *output_root != "") {
                     // Check if the path ends with a trailing slash,
                     // otherwise add it.
-                    return output_root->back() == '/'
+                    std::string str = output_root->back() == '/'
                            ? *output_root
                            : *output_root + "/";
-                }
 
+                    const char* dir = str.c_str();
+
+                    //use C++ system function to check if there is a dir match that defined in realization
+                    struct stat sb;
+                    if (stat(dir, &sb) == 0 && S_ISDIR(sb.st_mode))
+                        return dir;
+                    else {
+                        int result = mkdir(dir, 0755);      
+                        if (result == 0)
+                            return dir;
+                        else
+                            throw std::runtime_error("mkdir " + std::string(dir) + " failed, please create " + std::string(dir));
```

Comment:
Not critical, but I think we should make this error message more informational by checking what `errno` is set to (might need to include `<cerrno>`). Other than that, it looks okay.

```suggestion
                            throw std::runtime_error("failed to create directory '" + str + "': " + strerror(errno));
```

Reply:
Implemented as suggested.

Reply:
Oops, I think it actually should be `std::strerror`, my bad.